cassandra-quarkus-extension/deployment/src/main/java/com/datastax/oss/quarkus/deployment/CassandraClientProcessor.java

```diff
@@ -0,0 +1,171 @@
+  }
+
+  private String getCassandraClientProducerClassName() {
+    return AbstractCassandraClientProducer.class.getPackage().getName()
+        + "."
+        + "CassandraClientProducer";
+  }
+
+  private void createCassandraClientProducerBean(
+      BuildProducer<GeneratedBeanBuildItem> generatedBean,
+      String cassandraClientProducerClassName) {
+
+    ClassOutput classOutput = new GeneratedBeanGizmoAdaptor(generatedBean);
+
+    try (ClassCreator classCreator =
+        ClassCreator.builder()
+            .classOutput(classOutput)
+            .className(cassandraClientProducerClassName)
+            .superClass(AbstractCassandraClientProducer.class)
+            .build()) {
+      classCreator.addAnnotation(ApplicationScoped.class);
+
+      try (MethodCreator defaultCassandraClient =
+          classCreator.getMethodCreator("createDefaultCassandraClient", CqlSession.class)) {
+        defaultCassandraClient.addAnnotation(ApplicationScoped.class);
+        defaultCassandraClient.addAnnotation(Produces.class);
+        defaultCassandraClient.addAnnotation(Default.class);
+
+        // make CqlSession as Unremovable bean
+        defaultCassandraClient.addAnnotation(Unremovable.class);
+
+        ResultHandle cassandraClientConfig =
+            defaultCassandraClient.invokeVirtualMethod(
+                MethodDescriptor.ofMethod(
+                    AbstractCassandraClientProducer.class,
+                    "getCassandraClientConfig",
+                    CassandraClientConfig.class),
+                defaultCassandraClient.getThis());
+
+        defaultCassandraClient.returnValue(
+            defaultCassandraClient.invokeVirtualMethod(
+                MethodDescriptor.ofMethod(
+                    AbstractCassandraClientProducer.class,
+                    "createCassandraClient",
+                    CqlSession.class,
+                    CassandraClientConfig.class),
+                defaultCassandraClient.getThis(),
+                cassandraClientConfig));
+      }
+    }
+  }
+
+  @Record(RUNTIME_INIT)
+  @BuildStep
+  void configureRuntimePropertiesAndBuildClient(
+      CassandraClientRecorder recorder,
+      CassandraClientConfig cassandraConfig,
+      ConfigurationBuildItem config) {
+    recorder.configureRuntimeProperties(cassandraConfig);
+  }
+
+  @BuildStep
+  @Record(value = RUNTIME_INIT, optional = true)
+  CassandraClientBuildItem cassandraClient(CassandraClientRecorder recorder) {
+    return new CassandraClientBuildItem(recorder.getClient());
+  }
+
+  @BuildStep
+  HealthBuildItem addHealthCheck(CassandraClientBuildTimeConfig buildTimeConfig) {
+    return new HealthBuildItem(
+        "com.datastax.oss.quarkus.runtime.health.CassandraHealthCheck",
+        buildTimeConfig.healthEnabled,
+        "cassandra");
+  }
+
+  @BuildStep
+  RuntimeInitializedClassBuildItem runtimeMetadataManager() {
+    return new RuntimeInitializedClassBuildItem(MetadataManager.class.getCanonicalName());
+  }
```

Comment:
Why do we need to do this for `MetadataManager` specifically? It's just a regular internal driver class, there are many others like it.

Reply:
MetadataManager must be initialized at runtime because it uses Inet4Socket address that cannot be initialized at deployment time because of:
```
No instances of java.net.Inet4Address are allowed in the image heap as this class should be initialized at image runtime.
```
I think that this is the most narrow scope that we can make as the `RuntimeItem` but maybe we can make it even more narrow. I was trying to substitute `Inet4Address` but I was unable to do so.

Reply:
Would you mind explaining that in a comment? Otherwise, for reviewers it does look strange to see this component special-cased.

Reply:
good idea - done